DSD-350: Select refactor

bigfishdesign13 opened the pull request:
Fixes [DSD-350](https://jira.nypl.org/browse/DSD-350)

## **This PR does the following:**
- Refactors `Select` to match `TextInput` formatting
- Adds `errorText`, `helperText`, `labelText`, `required` and `showLabel` props to `Select`
- Deprecates `ariaLabel`, `helperTextId`, `isRequired` and `labelId` props in `Select`

### Front End Review:
- [ ] View [the example in Storybook]()

bigfishdesign13 committed:
initial development

bigfishdesign13 committed:
working on tests

bigfishdesign13 committed:
merge from development

bigfishdesign13 committed:
tests, changelog and polishing

EdwinGuzman commented on src/components/Select/Select.test.tsx:
```diff
@@ -51,42 +50,152 @@ describe("Select", () => {
+    expect(wrapper.find("select").prop("aria-label")).to.equal("Aria Label");
+  });
+
+  it("Renders aria-describedby", () => {
```
What makes it render an `aria-describedby`? Is it the `helperText` prop? Add that to this test `it` description, 
"Renders aria-describedby because a ____ prop was passed".

bigfishdesign13 replied:
Good suggestion.  Done.

EdwinGuzman commented on src/components/Select/Select.tsx:
```diff
@@ -13,14 +16,22 @@ export interface SelectProps {
+  isRequired?: boolean;
+  /** DEPRECATED - ID of associated label */
   labelId?: string;
+  /** Provides text for a `Label` component if `showLabel` is set to true; populates a `aria-label` sttribute if `showLabel` is set to false. */
```
Typo in `sttribute`.

bigfishdesign13 replied:
Done.

EdwinGuzman commented on src/components/Select/Select.tsx:
```diff
@@ -29,36 +40,61 @@ export interface SelectProps {
   onBlur?: (event: React.FormEvent) => void;
   /** Passes selects' current value to the React state handler */
   onChange?: (event: React.FormEvent) => void;
+  /** Will add 'aria-required: true' to input */
```
You should add that it also adds the native `required` attribute to the input (since browsers handle that differently).

bigfishdesign13 replied:
Done.

EdwinGuzman commented on src/components/Select/Select.tsx:
```diff
@@ -13,14 +16,22 @@ export interface SelectProps {
-  /** Attribute indicating that an option with a non-empty string value must be selected */
-  isRequired: boolean;
-  /** ID of associated label */
+  /** DEPRECATED - Attribute indicating that an option with a non-empty string value must be selected */
```
I think these labels are great to have, but maybe we can break the alphabetical list convention and put all the deprecated props at the bottom of the list and have one comment such as, "DEPRECATED - the following props are still available to use but are deprecated and will be removed in a future release", or similar.

EdwinGuzman replied:
This reminds me: maybe we can have a `console.warn` message if any of the deprecated props are used that says something like "These props used in the Select component will be deprecated in a future release: ...... Check the documentation for updated props to use instead."

bigfishdesign13 replied:
@EdwinGuzman I will add a `console.warn` message if/when a deprecated prop is used.  I like that a lot.  But for Storybook, what about hiding the props in the `ArgsTable` and including a note about the deprecated props below the ArgsTable?

EdwinGuzman replied:
Yes, hiding the props in Storybook and adding a note are good ideas.

bigfishdesign13 replied:
Done.

EdwinGuzman commented on src/components/Select/Select.tsx:
```diff
@@ -80,28 +116,41 @@ const Select = React.forwardRef<HTMLSelectElement, SelectProps>(
+          id={id}
+          className={bem("selectfield", modifiers, blockName, [className])}
+          aria-required={isRequired || required}
+          required={isRequired || required}
```
Nitpick: `isRequired || required` appears three times. Consider making a variable for this.

bigfishdesign13 replied:
Done.

EdwinGuzman commented on src/components/Select/_Select.scss:
```diff
@@ -2,47 +2,62 @@
+    appearance: none;
+    background-color: var(--ui-white);
+    /* Arrow svg */
+    /* More info here: https://css-tricks.com/probably-dont-base64-svg/ */
+    background-image: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' fill='currentColor' viewBox='0 0 24 24'%3E%3Cpath d='M10.526 12.871L.263 1.676 1.737.324 12 11.52 22.263.324l1.474 1.352L13.474 12.87a2 2 0 01-2.948 0z'/%3E%3C/svg%3E");
```
This is neat but mentions issues with some browsers that may no longer be a problem. Is there a benefit from doing this rather than rendering an `Icon` component?

bigfishdesign13 replied:
@EdwinGuzman This was how the component was initially put together.  Let me see what I can do.

bigfishdesign13 replied:
@EdwinGuzman Replacing the arrow on a `select` input is traditionally done with CSS.  I think refactoring the code to use the Icon component would be overkill.  That being said, having to copy and paste the `svg` text right into the CSS is burdensome.  In a webpack/react setup, how would I directly reference our arrow `svg`?

EdwinGuzman replied:
Ah gotcha. Let's leave this like this. It may be more complicated to get webpack involved than doing what's already here.

EdwinGuzman commented on src/components/Select/Select.stories.mdx:
```diff
@@ -0,0 +1,214 @@
+      errored: false,
+      required: true,
+      disabled: false,
+      selectedOption: "Option Two",
```
This isn't an option and leaving this here doesn't render an option that is selected, so it's better to remove this prop. This still shows up in the `Controls`.

bigfishdesign13 replied:
An oversight.  I changed the option values last minute and forgot to also adjust this prop.  Fixed now.

EdwinGuzman commented on src/components/Select/Select.stories.mdx:
```diff
@@ -0,0 +1,214 @@
+
+## Labelling Variations
+
+A Select can be rendered with or without visible labels.
```
Can you add that we do the job of adding an `aria-label` attribute when the label isn't visible so the Select component is still accessible?

bigfishdesign13 replied:
Done.

EdwinGuzman commented on src/components/Select/Select.stories.mdx:
```diff
@@ -0,0 +1,214 @@
+    <option aria-selected={false}>White</option>
+  </Select>
+</Canvas>
+
```
Can you add a preface here saying that this is how the `helperText` prop gets rendered?

EdwinGuzman replied:
Sorry, I meant this for the next example on line 127 when `showLabel` is set to true.

bigfishdesign13 replied:
I am not sure we need to explain the details behind the variations.  Adjusting the props in the ArgsTable should suffice.

EdwinGuzman commented on src/components/Select/Select.stories.mdx:
```diff
@@ -0,0 +1,214 @@
+    <option aria-selected={false}>White</option>
+  </Select>
+</Canvas>
+
```
Similar idea here, it'd be good to say that `errorText` and `errored` were passed for the next examples.

EdwinGuzman commented on src/components/Select/Select.stories.mdx:
```diff
@@ -0,0 +1,214 @@
+</Canvas>
+
+## Deprecated Implementation
+
```
Why is this a deprecated implementation? A short explanation would be good to have here.

bigfishdesign13 replied:
Done.

EdwinGuzman reviewed (changes requested):
Good so far but with some comments about adding more details. Also one question about rendering the arrow svg through css rather than through the `Icon` component.

bigfishdesign13 committed:
Merge branch 'development' into DSD-350-select-refactor

bigfishdesign13 committed:
addressing PR comments

bigfishdesign13 commented:
@EdwinGuzman I addressed all the comments/suggestions and this PR is ready for a second review, but github still has the PR listed as `Changes Requested`.

EdwinGuzman reviewed (approved):
Yeah, the "changes requested" goes away after approving.
